addons/project/tests/test_multicompany.py

```diff
@@ -220,7 +220,126 @@ def test_generate_analytic_account(self):
+        project_no_company.company_id = False
+        # set the company of the account
+        account_no_company.company_id = self.company_a
+        self.assertEqual(project_no_company.company_id, self.company_a, "The company of the project should have been updated to the company of its new account.")
+        self.assertEqual(account_no_company.company_id, self.company_a, "The company of the account should have been updated.")
+
+        # The project and its account have the same company (company A)
+        account_a.plan_id.company_id = False
+        # set the company of the project to False
+        self.project_company_a.company_id = False
+        self.assertFalse(self.project_company_a.company_id, "The company of the project should have been updated.")
+        self.assertFalse(account_a.company_id, "The company of the account should have been set to False, for its company was the same as the one of its project and the company of the project was set before the update.")
+        account_a.company_id = self.company_a
+        # set the company of the project to company B
+        self.project_company_a.company_id = self.company_b
+        self.assertEqual(self.project_company_a.company_id, self.company_b, "The company of the project should have been updated.")
+        self.assertEqual(account_a.company_id, self.company_b, "The company of the account should have been updated, for its company was the same as the one of its project and the company of the project was set before the update.")
+        # set the company of the account to company A
+        account_a.company_id = self.company_a
+        self.assertEqual(self.project_company_a.company_id, self.company_a, "The company of the project should have been updated to the company of its account.")
+        self.assertEqual(account_a.company_id, self.company_a, "The company of the account should have been updated.")
+        # set the company of the account to False
+        account_a.company_id = False
+        self.assertEqual(self.project_company_a.company_id, self.company_a, "The company of the project should not have been updated for the company of its account has been set to False.")
+        self.assertFalse(account_a.company_id, "The company of the account should have been updated.")
+
+        # The project has a company_id set, but not its account
+        # set the company of the account to company B (!= project.company_id)
+        account_a.company_id = self.company_b
+        self.assertEqual(self.project_company_a.company_id, self.company_b, "The company of the project should have been updated to the company of its account even if the new company set on the account is a different one than the one the project.")
+        self.assertEqual(account_a.company_id, self.company_b, "The company of the account should have been updated.")
```

Comment:
It is not the same use case :
For the line 289, because the project and its account have the same company, and we're updating the company of the account.
While on line 299 and 305, the project has a company set, but not its account. And we have to test both cases : one when the user set a company on the account, and that company is the same of its project, 
and when the user set a company on the account and that company is a different one from its project. 
But i agree with your points that the test are heavy and not very readable (i tried though), but the use case are quite precise, so we have to ensure that every use case is correctly handled in order to ensure that when something is updated, all the use case are correctly handled, and to ensure that future updates will not break some use cases. I'm not sure what's the best take there tbh. I prefer to have tests that covers the most use case we can think of, but on the other hand its gives very long and not very readable tests...